Make app responsive and toggle sidebar

sejas opened the pull request:
<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "Fixes" keyword and use "Related to" instead.
-->

- Closes https://github.com/Automattic/dotcom-forge/issues/8986

## Proposed Changes

- Create a new top bar
- Move Authentication buttons to that top bar
- Move offline indicator to the TopBar
- Add a new button to toggle the sidebar
- Fix tests (I'll move the topbar tests to its own file)

## Testing Instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

- Run `npm start`
- Observe a new top bar
- Click in the first icon
- Observe the app reduces its size and the sidebar disappears
- Change tabs and observe the UI doesn't break.
- Click again on the leftDrawer icon in the topbar and observe the app displays the sidebar and the window restores it's previous size.
- If possible test in Mac and Windows.


**Mac**


https://github.com/user-attachments/assets/ae3e928f-3ea6-4094-bb8d-bc9f5507ba60



<!-- https://github.com/user-attachments/assets/c9b53ad5-ea2e-4ac8-99a0-ba79cfad1991 !-->

**Windows**



https://github.com/user-attachments/assets/675cd76b-3f7b-40df-bb08-5c357e0d8c39



**Offline indicator**


<img width="1012" alt="Screenshot 2024-09-13 at 09 06 07" src="https://github.com/user-attachments/assets/48d7bc3d-66d3-4581-be5a-3ee835d5ce75">

<img width="1012" alt="Screenshot 2024-09-13 at 09 06 05" src="https://github.com/user-attachments/assets/e50f867b-a95e-4be3-9007-e03de8e56279">

## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] Have you checked for TypeScript, React or other console errors?

sejas committed:
Add rough topbar to toggle sidebar

sejas committed:
Resize window to set min width

sejas committed:
Add sidebar for Windows

sejas committed:
Add right spacing to respect WIndows window controls

sejas committed:
Move offline indicator to topbar

sejas committed:
Fix tests

sejas committed:
Avoid reducing the top bar when the main has scroll

sejas committed:
avoid displaying add site button over the main div

sejas committed:
display always at least two columns in overview shortcuts

sejas committed:
Move tests to topbar file

sejas committed:
Add test case to track Toggle Sidebar button

sejas commented:
@wojtekn , are you ok merging this? or would you prefer to add it behind a feature flag?

In my opinion the responsive is an independent feature from the Assistant, the same way the float on top is already released.

matt-west commented:
Nice work @sejas! I was only able to test it on Mac, but spotted a few minor things.

---

Is it possible to centre the icons in the top bar with the window controls?

<img width="1012" alt="Screenshot 2024-09-13 at 14 32 22" src="https://github.com/user-attachments/assets/65d10d20-a5ed-425b-81d8-039193f7af83">

 
Can you increase the height of the user element in the top bar to 24px so it matches the height of the help icon when hovered.

<img width="285" alt="Screenshot 2024-09-13 at 14 33 21" src="https://github.com/user-attachments/assets/4c421f83-37fb-4daf-9c49-511461e4098e">

The user avatar and help icon should be closer together. The avatar is also missing the white outline.

| Current | Design |
|--------|--------|
| <img width="217" alt="Screenshot 2024-09-13 at 14 40 46" src="https://github.com/user-attachments/assets/f1d33361-5f84-4917-b182-32f638d19994"> | <img width="230" alt="Screenshot 2024-09-13 at 14 41 30" src="https://github.com/user-attachments/assets/28e256b5-830a-409d-a65e-babb7968d088"> | 


When the offline icon is displayed, the header bar gets taller.

https://github.com/user-attachments/assets/2914bcc7-82b3-4dfd-8e62-4c6994d28289

Is it possible to automatically collapse the sidebar if the user tries to resize the window below the min-width? (Totally fine to tackle this in a separate PR if needed.)


https://github.com/user-attachments/assets/bd10f291-4b01-465c-8eea-f32af38414f9

wojtekn commented:
I've tested it on macOS. Looks great! I noticed some issues:
- layout shift between authenticated and logged-out state (mentioned by Matt above, too)
- The Import / Export tab got a scrollbar as we reduced window height. Could we slightly reduce the import box height to avoid that? @matt-west 
- when I have wide Studio window, and I hide the scrollbar, it hides the scrollbar but uses the minimum width. I expect it will use the previous width with the subtracted sidebar width

I will test it on Windows soon.

> In my opinion the responsive is an independent feature from the Assistant, the same way the float on top is already released.

It makes sense. Let's release it for all users.

matt-west commented:
> The Import / Export tab got a scrollbar as we reduced window height. Could we slightly reduce the import box height to avoid that?

Yep. Reducing the height to `h-36` should fix the scrolling issue.

wojtekn commented on src/components/windows-titlebar.tsx:
```diff
@@ -23,6 +29,7 @@ export default function WindowsTitlebar( { className }: { className?: string } )
 				<img src={ appIcon } alt="" className="w-[16px] flex-shrink-0" />
 				<h1 className="text-xs">{ getAppGlobals().appName }</h1>
 			</div>
+			<div className="flex-1 pl-2 pr-32	">{ children }</div>
```
Nitpick: not necessary space.

wojtekn commented:
It works well on Windows, with the following issues:

- minimum window width is used when I hide the scrollbar, regardless of my window width, similarly to macOS behavior
- when I hide the scrollbar, it is hidden smoothly with animation, but the window size changes rapidly. On macOS, it works smoothly, too. If it's not quick fix, we could fix this as a follow-up, though.

User logged in:

![Zrzut ekranu 2024-09-13 161108](https://github.com/user-attachments/assets/34b37020-2ec3-4b58-96c6-689836d9b12f)

Use logged out:
![Zrzut ekranu 2024-09-13 161125](https://github.com/user-attachments/assets/d917bc2c-89cb-4fba-8895-3a948d0ebfac)

Sidebar hidden:
![Zrzut ekranu 2024-09-13 161054](https://github.com/user-attachments/assets/2a888210-2692-43bc-9812-461befe8ebfa)

sejas committed:
Change animation speed

sejas committed:
center topbar with window buttons on MAC

sejas committed:
Increase gravatar size and reduce spacing in topbar

sejas committed:
Remove empty space in className

sejas committed:
Avoid jumping topbar when offline icon appears

sejas committed:
Fix login button height to avoid jumping when auth changes

sejas committed:
Reduce import area to avoid having a scroll bar in import export content tab

sejas committed:
remember both window sizes

sejas commented:
@wojtekn , @matt-west , thanks for reviewing the PR. I've fixed everything you mentioned.

> when I have wide Studio window, and I hide the scrollbar, it hides the scrollbar but uses the minimum width. I expect it will use the previous width with the subtracted sidebar width

I've implemented the memory size for both states, with and without sidebar, but this behaviour can conflict if/when we implement the window manual resizing:

> Is it possible to automatically collapse the sidebar if the user tries to resize the window below the min-width? (Totally fine to tackle this in a separate PR if needed.)

We could need a bit of discussion about making the window responsive by resizing it. It's possible, and it makes sense to me. But it seems incompatible combined  with saving the previous width.
My understanding of the manual resizing is that when a user makes the window smaller than the "minimum" size, the sidebar will disappear and when the user makes it bigger than the minimum size, then the sidebar will appear automatically. But what should happen when the user clicks on the button?. I think in that case we shouldn't save any previous width and just use the smallest width possible for each state.

I'll merge this PR and if we find new bugs or enhancements, we can open new PRs/issues. Thank you!

sejas commented:
@wojtekn , @matt-west , in fact Wojtek's suggestion was simpler and it's totally compatible to automatically displaying/hiding the sidebar on window resize. Check the PR below.

> when I have wide Studio window, and I hide the scrollbar, it hides the scrollbar but uses the minimum width. I expect it will use the previous width with the subtracted sidebar width

I misunderstood this explanation thinking to keep the size memory, but you actually were suggesting a simpler workflow. 👌 

I've created a PR applying your approach and Matt's suggestion.

> Is it possible to automatically collapse the sidebar if the user tries to resize the window below the min-width? (Totally fine to tackle this in a separate PR if needed.)

Done! 🥳  Let me know what you think:

- https://github.com/Automattic/studio/pull/536 😚 👌
